### ISSUE TITLE
Failure during retrieving images "Image as JSON string not supported"

### ISSUE THREAD

**Issue by TiPaKi** [kvos/CoastSat#87]
Dear Kilian, 

I planned to use Coastsat for my thesis, doing a temporal analyze of migrating sand spits in at the Baltic Sea coast. I succesfully created a coastsat environment, but unfortunately I am not able to retrieve the images from GEE. Neither with the example_jupyter nor with the Baltic Sea script. 
When I ran the "metadata = SDS_download.retrieve_images(inputs)" step it took hours and no result showed up. In both cases there were folders created in my current working directory, but those were empty. After several hours I interupted the kernel and got a fail report for the Baltic Sea script . For example_jupyter it was not possible to interupt. There was no reaction to the command, therefore also no fail report. I am working with Jupyter Notebook. This is my Baltic Sea script and the received fail report after interupting the kernel. The polygon has an area of ~ 2 square km. 

import os
import numpy as np
import pickle
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
from coastsat import SDS_download, SDS_preprocess, SDS_shoreline, SDS_tools, SDS_transects

polygon = [[[10.985512890687223, 54.37519400489867],
            [10.989546933045622, 54.3742691013092],
            [11.005768933167692, 54.37334417688454],
            [11.010661282410856,54.374544074822964],
            [11.025896229615446,54.37504402194819],
            [11.024737515321013,54.381642753479746],
            [10.996370472779509,54.380717995167224],
            [10.988516964783903,54.381592767076576],
            [10.98383919226193,54.3814678008023],
            [10.985512890687223,54.37519400489867]]]
              
dates = ['2019-01-01', '2019-01-31']

sat_list = ['L8']

sitename = 'Graswarder_Heiligenhafen'

filepath = os.path.join(os.getcwd(), 'graswarder_heiligenhafen_data_test3')

inputs = {'polygon': polygon, 'dates': dates, 'sat_list': sat_list, 'sitename': sitename, 'filepath':filepath}

SDS_download.check_images_available(inputs);

Images available between 2019-01-01 and 2019-01-31:
- In Landsat Tier 1 & Sentinel-2 Level-1C:
  L8: 2 images
  Total: 2 images
- In Landsat Tier 2:
  L8: 1 images
  Total: 1 images

metadata = SDS_download.retrieve_images(inputs)

Images available between 2019-01-01 and 2019-01-31:
- In Landsat Tier 1 & Sentinel-2 Level-1C:
  L8: 2 images
  Total: 2 images
- In Landsat Tier 2:
  L8: 1 images
  Total: 1 images

Downloading images:
L8: 2 images

EEException                               Traceback (most recent call last)
~\anaconda3\envs\coastsat\coastsat\SDS_download.py in retrieve_images(inputs)
    289                         im_ee = ee.Image(im_meta['id'])
--> 290                         local_data_pan = download_tif(im_ee, inputs['polygon'], bands['pan'], filepaths[1])
    291                         local_data_ms = download_tif(im_ee, inputs['polygon'], bands['ms'], filepaths[2])

~\anaconda3\envs\coastsat\coastsat\SDS_download.py in download_tif(image, polygon, bandsId, filepath)
    444 
--> 445     url = ee.data.makeDownloadUrl(ee.data.getDownloadId({
    446         'image': image.serialize(),

~\anaconda3\envs\coastsat\lib\site-packages\ee\data.py in getDownloadId(params)
    898     if isinstance(params['image'], six.string_types):
--> 899       raise ee_exception.EEException('Image as JSON string not supported.')
    900     params.setdefault('filePerBand', True)

EEException: Image as JSON string not supported.

During handling of the above exception, another exception occurred:

KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-4-890ef8172c0b> in <module>
      1 # inputs['include_T2'] = True
----> 2 metadata = SDS_download.retrieve_images(inputs)

~\anaconda3\envs\coastsat\coastsat\SDS_download.py in retrieve_images(inputs)
    288                     try:
    289                         im_ee = ee.Image(im_meta['id'])
--> 290                         local_data_pan = download_tif(im_ee, inputs['polygon'], bands['pan'], filepaths[1])
    291                         local_data_ms = download_tif(im_ee, inputs['polygon'], bands['ms'], filepaths[2])
    292                         break

KeyboardInterrupt: 

Have you seen this error before? I am just starting with this  Python and GEE stuff, therefore I am not very experienced until now and maybe thats why I dont see the problem. I would be thankfull if you have an idea how to solve this issue. 

Best wishes, 
Tina

**Comment by kvos** [comment]
Hi @TiPaKi , from looking at your script the way you defined the polygon is wrong and that is causing the error. You need to have 5 pairs of coordinates like in the example where the first and last point are the same ( only rectangles are allowed). Let me know if this fixes it

**Comment by TiPaKi** [comment]
Thanks for your fast reply @kvos.  I changed the coordinates. Can you give me a further hint how long it takes to retrieve the images for the Narra example? Its running several hours now and I am not sure wether this is realistic or there might be another problem. It does create new folders in my working directory but those are still empty. Thank you!

**Comment by kvos** [comment]
have you managed to run the Narrabeen example? it could be an underlying problem, so please try the Narrabeen example first so we can discard some sources of error. Then, how big is your polygon? if it's about 20-30 square kilometers it should download several images per minute

**Comment by TiPaKi** [comment]
I didnt manage to run the example. It stucks at the same cell as the script wit Baltic Sea coordinates. The polygon for that has a size of ~2 square km. I just tried to run the Narra example again. This is the script. I changed nothing.  

import os
import numpy as np
import pickle
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
from coastsat import SDS_download, SDS_preprocess, SDS_shoreline, SDS_tools, SDS_transects

polygon = [[[151.301454, -33.700754],
            [151.311453, -33.702075],
            [151.307237, -33.739761],
            [151.294220, -33.736329],
            [151.301454, -33.700754]]]

dates = ['2017-12-01', '2018-01-01']

sat_list = ['S2']

sitename = 'NARRA'

filepath_data = os.path.join(os.getcwd(), 'data')


inputs = {
    'polygon': polygon,
    'dates': dates,
    'sat_list': sat_list,
    'sitename': sitename,
    'filepath': filepath_data
        }


SDS_download.check_images_available(inputs);

metadata = SDS_download.retrieve_images(inputs)


The output after running SDS_download.retrieve_images(inputs) is following:

Images available between 2017-12-01 and 2018-01-01:
- In Landsat Tier 1 & Sentinel-2 Level-1C:
  S2: 7 images
  Total: 7 images

Downloading images:
S2: 7 images

So it seems to work quite well at first sight, but the downloading process does not come to an end. As I said, the data folder is created (including subfolders 10m, 20m, 60m and meta) but those stay empty. I also tried other working directories but it was resulting in the same issue.

**Comment by kvos** [comment]
ok so basically there is an infinite loop in the download code so that it doesn't break if your internet connection goes down or the gee server refuses you access. Have you tried disabling firewalls? because it seems like the first step works (you can access the gee server and see that there are 7 images available) but then you can't download the files. 
If that doesn't work, write `%pdb 1` and run again, it should print the error, please copy paste it here so I can have a look tomorrow

**Comment by TiPaKi** [comment]
The downloading procedere was the same when I deactivated the firewall. Debugging "metadata = SDS_download.retrieve_images(inputs)" led to the following output:


%debug
metadata = SDS_download.retrieve_images(inputs)

ERROR:root:No traceback has been produced, nothing to debug.
WARNING:googleapiclient.discovery_cache:file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\__init__.py", line 36, in autodetect
    from google.appengine.api import memcache
ModuleNotFoundError: No module named 'google.appengine'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\file_cache.py", line 33, in <module>
    from oauth2client.contrib.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\file_cache.py", line 37, in <module>
    from oauth2client.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\__init__.py", line 42, in autodetect
    from . import file_cache
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\file_cache.py", line 40, in <module>
    raise ImportError(
ImportError: file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
WARNING:googleapiclient.discovery_cache:file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\__init__.py", line 36, in autodetect
    from google.appengine.api import memcache
ModuleNotFoundError: No module named 'google.appengine'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\file_cache.py", line 33, in <module>
    from oauth2client.contrib.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\file_cache.py", line 37, in <module>
    from oauth2client.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\__init__.py", line 42, in autodetect
    from . import file_cache
  File "C:\Users\Tina\anaconda3\envs\coastsat\lib\site-packages\googleapiclient\discovery_cache\file_cache.py", line 40, in <module>
    raise ImportError(
ImportError: file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth

Images available between 2017-12-01 and 2018-01-01:
- In Landsat Tier 1 & Sentinel-2 Level-1C:
  S2: 7 images
  Total: 7 images

Downloading images:
S2: 7 images

Is the error caused by the missing modules "oauth2client" and "google.appengine"? I was not sure because the "Warning" says: "googleapiclient.discovery_cache:file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth" implying that I have a wrong version. But maybe its just a hint.

**Comment by kvos** [comment]
yes that's a good hint, it's a problem with connecting to the gee server. Can you type `conda list` in your environment and check that you have all the packages listed below installed.
```
google-api-core           1.14.2           py37h21ff451_0    conda-forge
google-api-python-client  1.7.11                     py_0    conda-forge
google-auth               1.6.3                      py_0    conda-forge
google-auth-httplib2      0.0.3                      py_2    conda-forge
google-cloud-core         1.0.3                      py_0    conda-forge
google-cloud-storage      1.20.0                     py_0    conda-forge
google-resumable-media    0.4.1                      py_0    conda-forge
googleapis-common-protos  1.6.0            py37h21ff451_0
```

**Comment by TiPaKi** [comment]
Looks like the packages are there. Here is a picture of it, so you can have a look. I have different versions than in your list od packages. Could this cause the error? 

![GEE Issue](https://user-images.githubusercontent.com/63994589/79942627-866cda00-8467-11ea-9fa5-538d358e40f5.PNG)

**Comment by kvos** [comment]
ok could be a version issue, try the following commands with your environment activated:
`conda clean --all`
`conda uninstall earthengine-api`
`conda install earthengine-api=0.1.173`

let me know if this fixes it

**Comment by TiPaKi** [comment]
It did. Download reaches 100% and finally there are images in the folder. Thank you so much for your help!

**Comment by kvos** [comment]
no worries tina, something must have gone wrong when you installed the environment and it went for another version, did you have any issues installing the environmnent or just followed the commands in the readme and all worked fine?

**Comment by TiPaKi** [comment]
I had to set up my environment twice. In both cases I followed the read me. The first environment was crashing at some point so that I was not even able to use python. I still dont know what the issue was, even after a lot of research. Then I deinstalled everything (including Python) and for the new one I imported the modules with Anaconda Navigator and not manually in Python as before to avoid that modules disturb each other because of version issues. I assumed that to be the reason for the first crash, but basically I am not sure. But as we saw, importing with the Anaconda Navigator seems not to be ideal for the coastsat environment.
